File: leetcode/add_binary/add_binary.cpp
/*
 * Time Complexity: O(N)
 *   - N is the length of the longer input string.
 *   - Each character is processed once, and the result is reversed at the end (also O(N)).
 *
 * Space Complexity: O(N)
 *   - The result string can be at most N+1 in length (for a possible carry).
 *   - Additional space for leading zeros is also O(N).
 */

#include<string>
#include<algorithm>

using std::string;

class Solution {
public:
    string addBinary(string a, string b) {
        string leadingZeros = "";
        int diffLength = abs(int(a.length() - b.length()));
        for(int i = 0; i < diffLength; i++) leadingZeros += '0';

        if(a.length() < b.length()) a = leadingZeros + a;
        else b = leadingZeros + b;
        
        int carry = 0, bitA, bitB, sum, bit;
        string result = "";

        for(int i = a.length()-1; i >= 0; i--) {
            bitA = a[i] - '0';
            bitB = b[i] - '0';
            sum = bitA + bitB + carry;
            bit = sum%2;
            carry = sum/2;
            result += (char)(bit + '0');
        }

        if(carry) {
            result += (char)(carry+'0');
        }

        reverse(begin(result), end(result));
        return result;
    }
};

int main() {
    Solution s;
    s.addBinary("11", "1");
}

File: leetcode/add_binary/add_binary.java
package add_binary;

class Solution {
    public String addBinary(String a, String b) {
        int diffLength = Math.abs(a.length() - b.length());
        String leadingZeros = "";
        for(int i = 0; i < diffLength; i++) leadingZeros += '0';
        
        if(a.length() < b.length()) a = leadingZeros + a;
        else b = leadingZeros + b;

        int carry = 0, bitA, bitB, bit, sum;
        StringBuilder ans = new StringBuilder();
        for(int i = a.length()-1; i >= 0; i--) {
            bitA = a.charAt(i)-'0';
            bitB = b.charAt(i)-'0';
            sum = bitA + bitB + carry;
            bit = sum%2;
            carry = sum/2;
            ans.append(bit);
        }

        if(carry > 0) {
            ans.append(carry);
        }

        return ans.reverse().toString();
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        s.addBinary("11", "1");
    }
}